src/triage/component/catwalk/storage.py

```diff
@@ -39,82 +47,88 @@ def exists(self):
         s3 = s3fs.S3FileSystem()
         return s3.exists(self.path)
 
-    def write(self, obj):
-        s3 = s3fs.S3FileSystem()
-        with s3.open(self.path, 'wb') as f:
-            joblib.dump(obj, f, compress=True)
-
-    def load(self):
-        s3 = s3fs.S3FileSystem()
-        with s3.open(self.path, 'rb') as f:
-            return joblib.load(f)
-
     def delete(self):
         s3 = s3fs.S3FileSystem()
         s3.rm(self.path)
 
+    @contextmanager
+    def open(self, *args, **kwargs):
+        s3 = s3fs.S3FileSystem()
+        with s3.open(self.path, *args, **kwargs) as f:
+            yield f
+
 
 class FSStore(Store):
     def exists(self):
         return os.path.isfile(self.path)
 
-    def write(self, obj):
-        os.makedirs(os.path.dirname(self.path), exist_ok=True)
-        with open(self.path, 'w+b') as f:
-            joblib.dump(obj, f, compress=True)
-
-    def load(self):
-        with open(self.path, 'rb') as f:
-            return joblib.load(f)
-
     def delete(self):
         os.remove(self.path)
 
+    @contextmanager
+    def open(self, *args, **kwargs):
+        os.makedirs(dirname(self.path), exist_ok=True)
+        with open(self.path, *args, **kwargs) as f:
+            yield f
+
 
 class MemoryStore(Store):
     store = None
 
     def exists(self):
         return self.store is not None
 
-    def write(self, obj):
-        self.store = obj
+    def write(self, bytestream):
+        self.store = bytestream
 
     def load(self):
         return self.store
 
     def delete(self):
         self.store = None
 
+    @contextmanager
+    def open(self, *args, **kwargs):
+        with BytesIO() as f:
+            yield f
+
+
+class ModelStore():
+    def __init__(self, store):
+        self.store = store
+
+    def write(self, obj):
+        with self.store.open('wb') as f:
+            joblib.dump(obj, f, compress=True)
+
+    def load(self):
+        with self.store.open('rb') as f:
+            return joblib.load(f)
+
+    def exists(self):
+        return self.store.exists()
+
+    def delete(self):
+        return self.store.delete()
+
 
 class ModelStorageEngine(object):
     def __init__(self, project_path):
         self.project_path = project_path
+        self.model_dir = 'trained_models'
 
     def get_store(self, model_hash):
         pass
 
 
 class S3ModelStorageEngine(ModelStorageEngine):
-    def __init__(self, *args, **kwargs):
-        super(S3ModelStorageEngine, self).__init__(*args, **kwargs)
-
     def get_store(self, model_hash):
-        full_path = os.path.join(self.project_path, 'trained_models', model_hash)
-        return S3Store(path=full_path)
+        return ModelStore(S3Store('/'.join([self.project_path, self.model_dir, model_hash])))
```

Comment:
I'm not sure how `S3Store` handles paths like `foo/bar//trained_models/123`, but it would seem safer, and likely necessary, to use a path utility:

    import posixpath
    posixpath.join('foo/', 'bar', …)

…or perhaps even safer:

    import pathlib
    pathlib.PurePosixPath('foo/', 'bar', …).as_posix()

…or the same:

    str(pathlib.PurePosixPath('foo/', 'bar', …))